core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractFileMetadata.java

```diff
@@ -0,0 +1,170 @@
+    String fullPath = fileStatus.getPath().toString();
+    String[] paths = fullPath.split("/");
+    this.fileName = paths[paths.length - 1];
+    this.timeStamp = fileStatus.getModificationTime();
+    this.owner = fileStatus.getOwner();
+    this.fileSize = fileStatus.getLen();
+    this.credentials = credentials;
+    this.permission = fileStatus.getPermission().toShort();
+
+    // generate file folder
+    this.fullPath = fullPath;
+
+    // check if file is a folder
+    this.isFolder = fileStatus.isDirectory();
+
+    // TODO: investigate how to cleanly set basePath
+    if (sourcePath.equals("")) {
+      basePath = this.fileName;
+    } else {
+      /*
+       * this block of code calculates how many folders (separated by "/")
+       * should be dropped from the prefix
+       */
+      int numSourcePaths = sourcePath.split("/").length;
+      int numStrip;
+      if (sourcePath.endsWith("/")) {
+        numStrip = numSourcePaths;
+        sourcePath = sourcePath.substring(0, sourcePath.length() - 1);
+      } else {
+        numStrip = numSourcePaths - 1;
+      }
+
+      /*
+       * this block of code drops the URI and reconstructs basePath by
+       * concatenating the folders with "/"
+       */
+      String pathWithoutURI = fullPath.substring(fullPath.indexOf(sourcePath));
+      String[] pathsWithoutURI = pathWithoutURI.split("/");
+      basePath = "";
+      for (int i = numStrip; i < pathsWithoutURI.length; i++) {
+        basePath = basePath.concat(pathsWithoutURI[i] + "/");
+      }
+      if (basePath.length() > 0) {
+        basePath = basePath.substring(0, basePath.length() - 1);
+      }
+    }
+  }
+
+  public AbstractFileMetadata(String fileName, String fullPath, long timeStamp, String owner,
+                              Long fileSize, Boolean isFolder, String basePath,
+                              short permission, Credentials credentials) {
+    this.fileName = fileName;
+    this.fullPath = fullPath;
+    this.timeStamp = timeStamp;
+    this.owner = owner;
+    this.fileSize = fileSize;
+    this.credentials = credentials;
+    this.isFolder = isFolder;
+    this.basePath = basePath;
+    this.permission = permission;
+  }
+
+  public String getFullPath() {
+    return fullPath;
+  }
+
+  public String getFileName() {
+    return fileName;
+  }
+
+  public long getFileSize() {
+    return fileSize;
+  }
+
+  public long getTimeStamp() {
+    return timeStamp;
+  }
+
+  public String getOwner() {
+    return owner;
+  }
+
+  public abstract Credentials getCredentials();
+
+  public Boolean getIsFolder() {
+    return this.isFolder;
+  }
+
+  public String getBasePath() {
+    return basePath;
+  }
+
+  public short getPermission() {
+    return permission;
+  }
+
+  /**
+   * extend from this class to create a credential class for each filesystem
+   */
+  public abstract static class Credentials {
+    public String databaseType;
```

Comment:
The information required to access the source filesystem will be stored in its subclass, so this class is called credentials.